src/terminal.c

```diff
@@ -465,6 +465,9 @@ ex_terminal(exarg_T *eap)
 	cmd = skipwhite(p);
     }
 
+    if (cmd == NULL || *cmd == NUL)
+	cmd = p_sh;
```

Comment:
No, it's ex_terminal() that's holding on to p_sh over a longer period of time (instead of dereferencing it just before use) that makes this bug possible.